File: Arrays/ContainsDuplicate.js
/* 
Given an integer array nums, return true if any value appears at least twice in the array, 
and return false if every element is distinct.

1) Brute force method
- Make a second array called AlreadyOccurs. Go through each element in given array
and check if it is in AlreadyOccurs. If it is, return true.

var used = [];
    for (var i = 0; i < nums.length; i++){
        if (used.includes(nums[i])){
            return true;
        }
        used[used.length] = nums[i];
    }
    return false;

2) Run Times
- Time: O(n^2)
- Space: O(n)

3) More efficient method
- Create a sum of past array valules. For each item, do modulo of sum. If == 0, then return true.
Time complexity is now O(n).

var arraySum = nums[0];
    for (var i = 1; i < nums.length; i++){
        if (arraySum % nums[i] == 0){
            return true;
        }
        arraySum += nums[i];
    }
    return false;

Didn't work.

Correct Answers: Sorting, Hashtable

*/

/**
 * @param {number[]} nums
 * @return {boolean}
 */
 var containsDuplicate = function(nums) {
    
};

File: Blind 75/Array/ContainsDuplicate.java
// 217. Contains Duplicate

class Solution {
    public boolean containsDuplicate(int[] nums) {
        HashMap<Integer, Integer> hmap = new HashMap<Integer, Integer>();
        for (int i=0; i<nums.length; i++){
            if (hmap.get(nums[i]) == null || hmap.get(nums[i]) != 1){
                hmap.put(nums[i], 1);
            } else {
                return true;
            }
        }
        return false;
    }
}

/*
Description: Given an integer arary, return true if any value appears at least twice in array.

Step 1 Brute Force: Create array of already visited nums. Go through nums and check if current num is in visited nums. 0(n)

Step 2: Create hashmap to keep track of visited nums. Key, Value = num, vistited (1) or not (2).

Completed in 5:30
*/